core/src/main/java/org/apache/iceberg/BaseSnapshot.java

```diff
@@ -150,6 +202,15 @@ private void cacheManifests() {
   }
 
   @Override
+  public List<ManifestFile> allManifests(FileIO fileIO) {
+    if (allManifests == null) {
+      cacheManifests(fileIO);
+    }
+    return allManifests;
+  }
+
+  @Override
+  @Deprecated
```

Comment:
For the SnapshotParser problem, I think we keep this un-deprecated and throw an error if we need to call `cacheManifests`.

Reply:
Agreed. The `SnapshotParser`'s usage of `FileIO` in `toJson` is really rare now (only if the manifest list location _isn't_ already in the snapshot, which was added quite some time ago).

And we don't have to necessarily change every deprecated call immediately to give people time to migrate, but if they have things that are that old, likely their work is relatively pretty custom anyway and they (or we) can find a way to ensure `cacheManifests` was already called in that case.

Reply:
Any suggestions on exception to throw if the changes are not already cached?